codec/src/main/java/io/netty/handler/codec/DefaultHeaders.java

```diff
@@ -211,54 +241,48 @@ public int size() {
 
     @Override
     public boolean isEmpty() {
-        return map.isEmpty();
+        return head == head.after;
     }
 
     @Override
     public Set<T> names() {
-        return unmodifiableSet(map.keySet());
+        Set<T> names = new LinkedHashSet<T>(size());
```

Comment:
I guess we could first check if its empty and if so just use Collections.emptySet() ?